docs/data/toolpad/core/components/crud/crud.md

```diff
@@ -0,0 +1,356 @@
+    // Fetch data from server
+    const data = await fetch('https://my-api.com/data', {
+      method: 'GET',
+      body: JSON.stringify({
+        page: paginationModel.page,
+        pageSize: paginationModel.pageSize,
+        sortModel,
+        filterModel,
+      }),
+    });
+
+    return data
+  },
+  // ...
+}
+```
+
+### `getOne`
+
+The `getOne` method in a data source can be used for retrieving a single specific item and displaying all of its attributes.
+
+This method must return an object corresponding to the item to be displayed.
+
+```tsx
+{
+  //...
+  getOne: (id) => {
+    // Fetch record from server
+    const record = await fetch(`https://my-api.com/data/${id}`, {
+      method: 'GET',
+    });
+
+    return record;
+  },
+  // ...
+}
+```
+
+### `createOne`
+
+The `createOne` method in a data source can be used for creating a new item, usually from a form where the value for each of its attributes can be specified.
+
+This method must return an object corresponding to the new item that has been created.
+
+```tsx
+{
+  //...
+  createOne: (data) => {
+    // Create record in the server
+    const record = await fetch('https://my-api.com/data', {
+      method: 'POST',
+      body: JSON.stringify(data),
+    });
+
+    return record;
+  },
+  // ...
+}
+```
+
+### `updateOne`
+
+The `updateOne` method in a data source can be used for updating an existing item, usually from a form where the value for each of its attributes can be specified.
+
+This method must return an object corresponding to the new item that has been updated.
+
+```tsx
+{
+  //...
+  updateOne: (id, data) => {
+    // Update record in the server
+    const record = await fetch(`https://my-api.com/data/${id}`, {
+      method: 'PUT',
+      body: JSON.stringify(data),
+    });
+
+    return record;
+  },
+  // ...
+}
+```
+
+### `deleteOne`
+
+The `deleteOne` method in a data source can be used for deleting an existing item.
+
+This method does not need to return anything.
+
+```tsx
+{
+  //...
+  deleteOne: (id) => {
+    // Delete record in the server
+    await fetch(`https://my-api.com/data/${id}`, {
+      method: 'DELETE',
+    });
+  },
+  // ...
+}
+```
+
+### Form validation
+
+Optionally, a `validate` function can be included in the data source in order to do form validation.
+
+This function takes an object with field values for a given item, and must return the corresponding errors for each field as strings, as shown in the example:
+
+```tsx
+{
+  //...
+  validate: (formValues) => {
+    const errors = {};
+
+    if (!formValues.name) {
+      errors.name = 'Name is required';
+    }
+    if (formValues.age < 18) {
+      errors.age = 'Age must be higher than 18';
+    }
+
+    return errors;
+  },
+  // ...
+}
+```
+
+This function will run automatically against the current values when the user tries to submit a form inside the `Crud` component, or changes any of its fields.
+
+#### Integration with external schema libraries
```

Comment:
Have you seen https://standardschema.dev/?

Reply:
Hadn't seen it, we can probably make the `validate` function be compatible with this standard.

Reply:
I've changed the format of the `validate` function to follow the Standard Schema, this should make it really straight forward to integrate with schema libraries that adhere to it: https://deploy-preview-4486--mui-toolpad-docs.netlify.app/toolpad/core/react-crud/#form-validation